doc/crdt-details.md

```diff
@@ -0,0 +1,1086 @@
+    };
+
+    // 9. move deleted or undone-inserted things from text to tombstones
+    let (new_text, new_tombstones) = shuffle(&text_with_inserts, &self.tombstones,
+        &rebased_deletes_from_union, &new_deletes_from_union);
+
+    let head_rev = &self.revs.last().unwrap();
+    (Revision {
+        rev_id: self.next_rev_id(),
+        max_undo_so_far: std::cmp::max(undo_group, head_rev.max_undo_so_far),
+        edit: Edit {
+            priority: new_priority,
+            undo_group: undo_group,
+            inserts: new_inserts,
+            deletes: new_deletes,
+        }
+    }, new_text, new_tombstones, new_deletes_from_union)
+}
+
+// TODO: have `base_rev` be an index so that it can be used maximally efficiently with the
+// head revision, a token or a revision ID. Efficiency loss of token is negligible but unfortunate.
+pub fn edit_rev(&mut self, priority: usize, undo_group: usize,
+        base_rev: RevToken, delta: Delta<RopeInfo>) {
+    let (new_rev, new_text, new_tombstones, new_deletes_from_union) =
+        self.mk_new_rev(priority, undo_group, base_rev, delta);
+    self.rev_id_counter += 1;
+    self.revs.push(new_rev);
+    self.text = new_text;
+    self.tombstones = new_tombstones;
+    self.deletes_from_union = new_deletes_from_union;
+}
+```
+
+#### InsertDelta::transform_expand
+
+This is another big function with lots of nested loops. Unlike `Subset::transform_expand` it takes a boolean parameter `after` that resolves the ambiguity of whether inserts should go before or after the characters being transformed over.
+
+```rust
+/// Do a coordinate transformation on an insert-only delta. The `after` parameter
+/// controls whether the insertions in `self` come after those specific in the
+/// coordinate transform.
+pub fn transform_expand(&self, xform: &Subset, after: bool) -> InsertDelta<N> {
+    // ...
+}
+```
+
+#### shuffle
+
+This helper is used in many different operations that use the pattern of computing how `deletes_from_union` should change and then using that to fix up the `text` and `tombstones` to correspond to that.
+
+```rust
+/// Move sections from text to tombstones and vice versa based on a new and old set of deletions.
+/// Returns a tuple of a new text `Rope` and a new `Tombstones` rope described by `new_deletes_from_union`.
+fn shuffle(text: &Rope, tombstones: &Rope,
+        old_deletes_from_union: &Subset, new_deletes_from_union: &Subset) -> (Rope,Rope) {
+    // Delta that deletes the right bits from the text
+    let del_delta = Delta::synthesize(tombstones, old_deletes_from_union, new_deletes_from_union);
+    let new_text = del_delta.apply(text);
+    (new_text, shuffle_tombstones(text,tombstones,old_deletes_from_union,new_deletes_from_union))
+}
+```
+
+### Engine::undo
+
+Undo works conceptually by rewinding to the earliest point the toggled undo group appears, and replaying history from there but with the correct undone groups not applied.
```

Comment:
'correct' is ambiguous here. Maybe "[...] replaying history, skipping toggled undo groups."?